packages/amplify_core/lib/src/types/common/amplify_auth_provider.dart

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:aws_signature_v4/aws_signature_v4.dart';
+
+/// An identifier to use as a key in an [AmplifyAuthProviderRepository] so that
+/// a retrieved auth provider can be typed more accurately.
+class AmplifyAuthProviderToken<T extends AmplifyAuthProvider> extends Token<T> {
+  const AmplifyAuthProviderToken();
+}
+
+abstract class AuthProviderOptions {
+  const AuthProviderOptions();
+}
+
+/// Options required by IAM to sign any given request at runtime.
+class IamAuthProviderOptions extends AuthProviderOptions {
+  final String region;
+  final AWSService service;
+
+  const IamAuthProviderOptions({required this.region, required this.service});
+}
+
+class ApiKeyAuthProviderOptions extends AuthProviderOptions {
+  final String apiKey;
+
+  const ApiKeyAuthProviderOptions(this.apiKey);
+}
+
+abstract class AmplifyAuthProvider {
+  Future<AWSBaseHttpRequest> authorizeRequest(
+    AWSBaseHttpRequest request, {
+    covariant AuthProviderOptions? options,
+  });
+}
+
+abstract class AWSIamAmplifyAuthProvider extends AmplifyAuthProvider
+    implements AWSCredentialsProvider {
+  @override
+  Future<AWSSignedRequest> authorizeRequest(
+    AWSBaseHttpRequest request, {
+    covariant IamAuthProviderOptions options,
+  });
+}
+
+abstract class ApiKeyAmplifyAuthProvider extends AmplifyAuthProvider {
+  @override
+  Future<AWSBaseHttpRequest> authorizeRequest(
+    AWSBaseHttpRequest request, {
+    covariant ApiKeyAuthProviderOptions? options,
+  });
+}
+
+abstract class TokenAmplifyAuthProvider extends AmplifyAuthProvider {
+  Future<String> getLatestAuthToken();
+  Future<String> getIdentityId();
```

Comment:
suggestion: instead of adding this to this class, could you add a new abstract class that extends this one called maybe `TokenIdentityAmplifyAuthProvider` and add this method there?  Then, in auth plugin, make the cognito provider implement this new one? The reason is because in the case of OIDC/lambda auth mode in API, there will be a token provider that provides a token but not an identity ID so it will be nice to have a class without the identityId. Alternatively, that case could be no-op for this but I don't think that's correct.

I realize I suggested what you have here last time we spoke about this my bad for contradiction, was just thinking about it more.